### ISSUE TITLE
Add "/fleetcarrier" cAPI endpoint data to EDMC plugin's API

### ISSUE THREAD

**Issue by norohind** [EDCD/EDMarketConnector#847]
**Describe the enhancement**

Add information from https://companion.orerve.net/fleetcarrier to EDMC plugin's API as it done with "/profile" endpoint and "cmdr_data" function. 

**Expected behavior**

Call function with response of endpoint and beta state as it done with "cmdr_data" function.

**Considered workarounds**

It is possible to get current fleet carrier location by checking where your ships stored (by data from /profile endpoint). But it requires to store on carrier at least 1 ship. Also it doesn't provide any other information about carrier, just location.

**Additional context**
I can't imagine special conditions for automatically send request to this endpoint so i think there should be manual way to fetch data from this endpoint.

**Comment by Athanasius** [comment]
Some notes for implementation:

1. We don't want to just sling out `/fleetcarrier` requests for every user.  The vast majority likely won't own an FC.
    1. There doesn't seem to be any indication in the Journal that the Cmdr owns an FC until one of the relevant events occurs, so we can't tell on Cmdr login and set a flag.
    2. Events to know we should start querying CAPI `/fleetcarrier` for the current Cmdr are going to be *any* of the events in section 11 of v28 Journal docs.
    3. We'll need to have a flag (might just be able to use a stored FC callsign) that we set when we first detect the Cmdr has an FC, and unset/set it False when we confirm a Decommission.  NB: There's no event for a Decommission completing.  This will need inferring from the CarrierDecommission event timestamp, plus the known time it takes, and/or when `/fleetcarrier` no longer returns valid data.
2. When would we actually fire off `/fleetcarrier` ?
    1. We'll want to do so after seeing relevant FC Journal events. However this probably needs to go through a 'queue' system so that we do *not* request `/fleetcarrier` for every event, which could mean several a minute.  So start a timer when we see one (and update this for each new event), when that expires, send the request.  Some care will be required in selecting this timer length to avoid the request *never* being sent for a 'fussy' Cmdr.
    2. In addition to this we'll want to periodically poll as things like the current status of Cargo can change without the Cmdr themselves causing the change.

**Comment by Athanasius** [comment]
Re: point 1 - actually we can afford to sling out a single `/fleetcarrier` after initial authorization and then (re)set the flag based on that.  We can then later change it if we detect an FC being bought.

**Comment by Athanasius** [comment]
NB: Apparently a `/fleetcarrier` request can take 20 seconds or more, due to the in-FC cargo being returned as one entry per T of cargo.

1. The request absolutely must be in its own thread.
2. We probably want a timeout of 60s.
3. We won't want to query more often than somethng like 180s.

**Comment by Athanasius** [comment]
You know what?  Might as well have a config setting to control if we even attempt these queries.  Default to off obviously.  That will save a bunch of load on the CAPI servers.

**Comment by A-UNDERSCORE-D** [comment]
We could also maybe ping once every 12 hours or so to ask if one exists? then do something based on that?

**Comment by norohind** [comment]
In journals we have "CarrierBuy" event that can be used to detect new one.

**Comment by Athanasius** [comment]
I now have my own Fleet Carrier so could actually do easy direct testing of any related code.

Don't hold your breath though.  We need to get things settled down after Odyssey launch first, at least.

**Comment by Athanasius** [comment]
It isn't literally one cargo entry per 1T, but it does look to be an entry per transaction that got the cargo there.  i.e. I have several 728T Basic Medicines entries because that's max cargo on my Imperial Cutter, but also some 1T ones.

We're absolutely going to need *first* to make all CAPI queries asynchronous, which will be a fairly major re-engineering.

**Comment by Athanasius** [comment]
Example CAPI `/fleetcarrier` output for reference:

[Athanasius-fleetcarrier-20210808.json.txt](https://github.com/EDCD/EDMarketConnector/files/6951456/Athanasius-fleetcarrier-20210808.json.txt)

**Comment by Athanasius** [comment]
This is blocked until #945 is complete.

**Comment by Athanasius** [comment]
It has been brought to my attention that there's an extra caveat with the `/fleetcarrier` endpoint.  The data it returns can easily be 15-20 minutes out of date, if not 30 minutes or more, even if there have been changes on the FC itself that would cause a change to the data.

It's presumed this is Frontier either actively enforcing a cache of the data, or technical reasons meaning it exists.

So, sure, we can still implement this, but it's starting to look less and less useful.

**Comment by Athanasius** [comment]
Another point to note for coding this.

If a `/fleetcarrier` query is going to take a long time then it's feasible for several 'other' queries to be queued up in the meantime.  When a new query is queued it should cause discarding of any older duplicates.  Thus once things are unblocked we'll only trigger *one* 'other' query, not multiple.

**Comment by Athanasius** [comment]
Closing as the implementation is now merged into `develop`.
